diff -u freetype-2.4.11/include/freetype/config/ftconfig.h.orig freetype-2.4.11/include/freetype/config/ftconfig.h
--- freetype-2.4.11/include/freetype/config/ftconfig.h.orig
+++ freetype-2.4.11/include/freetype/config/ftconfig.h
@@ -483,10 +483,14 @@
 
 #ifndef FT_EXPORT
 
 #if defined(_WIN32) && defined(FT2_BUILD_LIBRARY) && defined(DLL_EXPORT)
 #define FT_EXPORT( x )  __declspec(dllexport)  x
+
+#elif defined(HAVE_VISIBILITY_ATTR) && defined(FT2_BUILD_LIBRARY)
+#define FT_EXPORT( x )  __attribute__(( visibility( "default" ) ))  x
+
 #else
 
 #ifdef __cplusplus
 #define FT_EXPORT( x )  extern "C"  x
 #else
diff -u /dev/null freetype-2.4.11/Makefile.darwin
--- /dev/null
+++ freetype-2.4.11/Makefile.darwin
@@ -0,0 +1,80 @@
+CROSS	=
+CC	= $(CROSS)gcc
+
+DYLIB = libfreetype.6.dylib
+
+INCPATH =-Iinclude
+CFLAGS = -O2 -Wall -fno-common
+CFLAGS+= -DFT2_BUILD_LIBRARY -DDARWIN_NO_CARBON -DNDEBUG
+CFLAGS+= -DFT_CONFIG_OPTION_SYSTEM_ZLIB -DFT_CONFIG_OPTION_USE_BZIP2
+CFLAGS+= -DHAVE_VISIBILITY_ATTR -fvisibility=hidden
+CFLAGS+= $(INCPATH)
+LDFLAGS= -dynamiclib -Wl,-single_module -Wl,-undefined,error -Wl,-install_name,"@rpath/FreeType.framework/Versions/A/FreeType" -Wl,-compatibility_version,15.0 -Wl,-current_version,17.0
+LDLIBS = -lz -lbz2
+#specifically for 10.5+:
+CFLAGS+=-mmacosx-version-min=10.5
+LDFLAGS+=-mmacosx-version-min=10.5
+
+SRCS =	src/autofit/autofit.c \
+	src/base/ftbase.c \
+	src/base/ftbbox.c \
+	src/base/ftbdf.c \
+	src/base/ftbitmap.c \
+	src/base/ftcid.c \
+	src/base/ftfstype.c \
+	src/base/ftgasp.c \
+	src/base/ftglyph.c \
+	src/base/ftgxval.c \
+	src/base/ftinit.c \
+	src/base/ftmm.c \
+	src/base/ftotval.c \
+	src/base/ftpatent.c \
+	src/base/ftpfr.c \
+	src/base/ftstroke.c \
+	src/base/ftsynth.c \
+	src/base/fttype1.c \
+	src/base/ftxf86.c \
+	src/base/ftwinfnt.c \
+	src/bdf/bdf.c \
+	src/cache/ftcache.c \
+	src/cff/cff.c \
+	src/cid/type1cid.c \
+	src/gzip/ftgzip.c \
+	src/bzip2/ftbzip2.c \
+	src/lzw/ftlzw.c \
+	src/pcf/pcf.c \
+	src/pfr/pfr.c \
+	src/psaux/psaux.c \
+	src/pshinter/pshinter.c \
+	src/psnames/psnames.c \
+	src/raster/raster.c \
+	src/sfnt/sfnt.c \
+	src/smooth/smooth.c \
+	src/truetype/truetype.c \
+	src/type1/type1.c \
+	src/type42/type42.c \
+	src/winfonts/winfnt.c \
+	builds/unix/ftsystem.c \
+	src/base/ftdebug.c
+
+OBJ = $(SRCS:.c=.o)
+
+all: $(DYLIB)
+
+.PHONY: clean distclean
+.SUFFIXES:
+.SUFFIXES: .o .c
+
+.c.o:
+	$(CC) -c $(CFLAGS) -o $@ $<
+
+builds/unix/ftsystem.o: builds/unix/ftsystem.c
+	$(CC) -c $(CFLAGS) -DHAVE_UNISTD_H -DHAVE_FCNTL_H -DHAVE_STDINT_H -Iinclude/freetype/config -o $@ $<
+
+$(DYLIB): $(OBJ)
+	$(CC) -o $(DYLIB) $(OBJ) $(LDFLAGS) $(LDLIBS)
+
+distclean: clean
+	$(RM) *.dylib
+clean:
+	$(RM) $(OBJ)
